tests/scripts/all.sh

```diff
@@ -3934,27 +3934,58 @@ component_build_tfm() {
     make lib CC="gcc" CFLAGS="-Os -std=c99 -Werror -Wall -Wextra -Wwrite-strings -Wpointer-arith -Wshadow -Wvla -Wformat=2 -Wno-format-nonliteral -Wshadow -Wformat-signedness -Wlogical-op -I../tests/include/spe"
 }
 
-component_build_aes_variations() { # ~45s
+component_build_aes_variations() { # ~7m
     # aes.o has many #if defined(...) guards that intersect in complex ways.
     # Test that all the combinations build cleanly. The most common issue is
     # unused variables/functions, so ensure -Wunused is set.
 
     msg "build: aes.o for all combinations of relevant config options"
 
+    # check to see if we can enable MBEDTLS_AES_USE_HARDWARE_ONLY - require
+    # Linux (so we can check for CPU flags)
+    if [[ "$OSTYPE" == "linux-gnu" ]]; then
+        # Runtime detection is supported on Linux, so it's safe to set these here
+        AESNI_OPTIONS="set unset"
+        AESCE_OPTIONS="set unset"
+    else
+        # otherwise leave them unset
+        AESNI_OPTIONS="unset"
+        AESCE_OPTIONS="unset"
+    fi
+
     for a in set unset; do
     for b in set unset; do
     for c in set unset; do
     for d in set unset; do
     for e in set unset; do
     for f in set unset; do
     for g in set unset; do
+    for h in set unset; do
+    for i in ${AESNI_OPTIONS}; do
+    for j in ${AESCE_OPTIONS}; do
+    for k in set unset; do
+        if [[ "$h" == "set" ]]; then
+            if [[ !(("$HOSTTYPE" == "aarch64" && "$j" == "set") || ("$HOSTTYPE" == "x86_64" && "$i" == "set")) ]]; then
+                # MBEDTLS_AES_USE_HARDWARE_ONLY requires hw acceleration for the target platform
+                continue
+            fi
+            if [[ "$g" == "set" ]]; then
+                # MBEDTLS_AES_USE_HARDWARE_ONLY and MBEDTLS_PADLOCK_C is not supported
+                continue
+            fi
+        fi
+
         echo ./scripts/config.py $a MBEDTLS_AES_SETKEY_ENC_ALT
         echo ./scripts/config.py $b MBEDTLS_AES_DECRYPT_ALT
         echo ./scripts/config.py $c MBEDTLS_AES_ROM_TABLES
         echo ./scripts/config.py $d MBEDTLS_AES_ENCRYPT_ALT
         echo ./scripts/config.py $e MBEDTLS_AES_SETKEY_DEC_ALT
         echo ./scripts/config.py $f MBEDTLS_AES_FEWER_TABLES
         echo ./scripts/config.py $g MBEDTLS_PADLOCK_C
+        echo ./scripts/config.py $h MBEDTLS_AES_USE_HARDWARE_ONLY
+        echo ./scripts/config.py $i MBEDTLS_AESNI_C
+        echo ./scripts/config.py $j MBEDTLS_AESCE_C
+        echo ./scripts/config.py $k MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH
```

Comment:
If you (Tom, whoever) is looking to optimize, `config.py` is not particularly fast. Adding a mode that allows you set/unset a bunch of options in a single invocation might make a significant difference here.

The reason there's no such option is that `config.py` is backward-compatible with its ancestor `config.pl`. The perl code was faster, but much less flexible.

Reply:
Or even faster: unset all of these options before the loop, then use `-D` on the compiler command line.

Also, you could invoke `clang` directly instead of running `make`: you're only building a single file, in a controlled environment.

Reply:
Yes, I was experimenting locally with using `sed` instead. That, plus skipping make and using `clang -fsyntax-only` gets it from 7m down to 1:52, which is worth having.

Reply:
Actually, if `config.py` is slow, just moving the loops that might-not-be-run further out from innermost would save time